src/adapter/src/catalog/storage/objects.rs

```diff
@@ -7,249 +7,706 @@
+                az_user_specified,
+            } => proto::replica_config::Location::Managed(proto::replica_config::ManagedLocation {
+                size: size.to_string(),
+                availability_zone: availability_zone.to_string(),
+                az_user_specified: *az_user_specified,
+            }),
+        }
+    }
+
+    fn from_proto(proto: proto::replica_config::Location) -> Result<Self, TryFromProtoError> {
+        match proto {
+            proto::replica_config::Location::Unmanaged(location) => {
+                Ok(SerializedReplicaLocation::Unmanaged {
+                    storagectl_addrs: location.storagectl_addrs,
+                    storage_addrs: location.storage_addrs,
+                    computectl_addrs: location.computectl_addrs,
+                    compute_addrs: location.compute_addrs,
+                    workers: CastFrom::cast_from(location.workers),
+                })
+            }
+            proto::replica_config::Location::Managed(location) => {
+                Ok(SerializedReplicaLocation::Managed {
+                    size: location.size,
+                    availability_zone: location.availability_zone,
+                    az_user_specified: location.az_user_specified,
+                })
+            }
+        }
+    }
+}
 
-    fn into_stash(self) -> Self::Stash {
-        let value = match self {
-            RoleId::User(id) => proto::role_id::Value::User(id),
-            RoleId::System(id) => proto::role_id::Value::System(id),
-            RoleId::Public => proto::role_id::Value::Public(Default::default()),
+impl RustType<proto::DatabaseKey> for DatabaseKey {
+    fn into_proto(&self) -> proto::DatabaseKey {
+        let value = match self.ns {
+            None | Some(DatabaseNamespace::User) => proto::database_id::Value::User(self.id),
+            Some(DatabaseNamespace::System) => proto::database_id::Value::System(self.id),
         };
 
-        proto::RoleId { value: Some(value) }
+        proto::DatabaseKey {
+            id: Some(proto::DatabaseId { value: Some(value) }),
+        }
     }
 
-    fn from_stash(stash: Self::Stash) -> Result<Self, anyhow::Error> {
-        match stash.value {
-            Some(proto::role_id::Value::User(id)) => Ok(RoleId::User(id)),
-            Some(proto::role_id::Value::System(id)) => Ok(RoleId::System(id)),
-            Some(proto::role_id::Value::Public(_)) => Ok(RoleId::Public),
-            None => Err(anyhow::anyhow!("Invalid RoleId, found None value")),
+    fn from_proto(proto: proto::DatabaseKey) -> Result<Self, TryFromProtoError> {
+        let id = proto
+            .id
+            .ok_or_else(|| TryFromProtoError::missing_field("DatabaseKey::id"))?;
+        match id.value {
+            Some(proto::database_id::Value::User(id)) => Ok(DatabaseKey {
+                id,
+                ns: Some(DatabaseNamespace::User),
+            }),
+            Some(proto::database_id::Value::System(id)) => Ok(DatabaseKey {
```

Comment:
Might want to test this

Reply:
Currently there isn't a way to create a "system database". I'm thinking about adding some more randomized testing with proptest, do you mind if I come back to this case?

Reply:
No, that's fine. I just thought it might be interesting for you to know which parts are not covered by existing tests.